core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,259 @@
+        @Nullable
+        private volatile Subscription upstream;
+
+        private volatile long requested;
+        private volatile long needToPublish;
+        private volatile boolean subscribed;
+        private volatile boolean cancelled;
+
+        SurroundingSubscriber(@Nullable T head, Publisher<? extends T> publisher, @Nullable T tail,
+                              Subscriber<? super T> downstream) {
+            requireNonNull(publisher, "publisher");
+            requireNonNull(downstream, "downstream");
+            state = head != null ? State.REQUIRE_HEAD : State.REQUIRE_BODY;
+            this.head = head;
+            this.publisher = publisher;
+            this.tail = tail;
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (cancelled || state == State.DONE) {
+                return;
+            }
+            for (;;) {
+                final long oldRequested = requested;
+                final long newRequested = LongMath.saturatedAdd(oldRequested, n);
+                if (requestedUpdater.compareAndSet(this, oldRequested, newRequested)) {
+                    if (oldRequested > 0) {
+                        return;
+                    }
+                    break;
+                }
+            }
+
+            publish();
+        }
+
+        private void publish() {
+            for (;;) {
+                if (requested <= 0) {
+                    return;
+                }
+                switch (state) {
+                    case REQUIRE_HEAD: {
+                        sendHead();
+                        continue;
+                    }
+                    case REQUIRE_BODY: {
+                        if (!subscribed) {
+                            subscribed = true;
+                            publisher.subscribe(this);
+                            return;
+                        }
+                        if (upstream != null) {
+                           requestUpstream(upstream);
+                        }
+                        return;
+                    }
+                    case REQUIRE_TAIL: {
+                        sendTail();
+                        continue;
+                    }
+                    case REQUIRE_COMPLETE: {
+                        sendComplete();
+                        return;
+                    }
+                    case DONE: {
+                        upstream.cancel();
+                        return;
+                    }
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            requestedUpdater.decrementAndGet(this);
+            downstream.onNext(head);
```

Comment:
`requestedUpdater.decrementAndGet(this)` needs to be decreased after `downstream.onNext(head)`. Otherwise, the order of elements could be mixed if `requested` becomes 0 after decrement.
```suggestion
            downstream.onNext(head);
            requestedUpdater.decrementAndGet(this);
```

Reply:
✅ updated! thanks for catching 🙇